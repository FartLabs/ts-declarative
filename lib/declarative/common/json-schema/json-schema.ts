import { TypeBoxFromSyntax } from "@sinclair/typemap";
import type { Declarative } from "#/lib/declarative/declarative.ts";
import type { ValueTsMorph } from "#/lib/declarative/common/ts-morph/ts-morph.ts";

export interface ValueJSONSchema extends ValueTsMorph {
  // deno-lint-ignore no-explicit-any
  jsonSchema?: any;
}

export function declarativeJSONSchema<
  TValue extends ValueJSONSchema,
>(): Declarative<TValue> {
  return (value) => {
    return { ...value, jsonSchema: compile(value) };
  };
}

export function compile({ tsMorph }: ValueTsMorph) {
  return TypeBoxFromSyntax({}, serialize({ tsMorph }));
}

export function serialize({ tsMorph }: ValueTsMorph): string {
  if (tsMorph === undefined || tsMorph?.properties.length === 0) {
    return `{}`;
  }

  return `{ ${
    tsMorph.properties
      .map((property) => `"${property.name}": ${property.type}`)
      .join(", ")
  } }`;
}
