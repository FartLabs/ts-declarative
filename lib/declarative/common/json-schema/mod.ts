export * from "./json-schema.ts";
