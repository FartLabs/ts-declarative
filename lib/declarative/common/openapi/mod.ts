export * from "./paths-object.ts";
export * from "./specification.ts";
export * from "./server.ts";
