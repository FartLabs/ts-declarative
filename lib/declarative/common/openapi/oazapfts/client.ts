import type { Opts } from "oazapfts";
import { generateSource } from "oazapfts";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import type { ValueOpenAPISpecification } from "#/lib/declarative/common/openapi/specification.ts";

/**
 * generateOazapftsClientOf returns the source code generated by Oazapfts.
 *
 * @example
 * ```ts
 * @standardCreate()
 * @standardGet()
 * @autoSchema()
 * class Person {
 *   public constructor(public name: string) {}
 * }
 *
 * @openapi({ resources: [Person] })
 * class App {}
 *
 * const sourceCode = await generateOazapftsClientOf(App);
 * await Deno.writeTextFile("client.ts", sourceCode);
 * ```
 */
export async function generateOazapftsClientOf(
  target: Class,
  options?: OazapftsOptions,
): Promise<string | undefined> {
  return await generateSource(
    getPrototypeValue<ValueOpenAPISpecification>(target)
      ?.specification as unknown as string,
    options,
  );
}

/**
 * OazapftsOptions is the options for the Oazapfts decorator.
 */
export type OazapftsOptions = Opts;
