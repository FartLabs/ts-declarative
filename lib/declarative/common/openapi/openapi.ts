import type { Route } from "@std/http/unstable-route";
import type { OpenAPIV3_1 } from "openapi-types";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";

/**
 * specificationOf returns the OpenAPI specification of the class.
 */
export function specificationOf<TClass extends Class>(
  target: TClass,
): OpenAPIV3_1.Document | undefined {
  return getPrototypeValue<ValueOpenAPI>(target)?.specification;
}

/**
 * routesOf returns the HTTP routes of the OpenAPI class.
 */
export function routesOf<TClass extends Class>(
  target: TClass,
): Route[] | undefined {
  return getPrototypeValue<ValueOpenAPI>(target)?.routes;
}

export function openapiDecoratorFactory(): () => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (value?: ValueOpenAPI) => {
      return [declarativeOpenAPI(value)];
    },
  });
}

export function declarativeOpenAPI<TValue extends ValueOpenAPI>(
  value0?: TValue,
): Declarative<TValue> {
  return (value1) => {
    return Object.assign({}, value0, value1);
  };
}

/**
 * ValueOpenAPI is the value associated with an OpenAPI specification.
 */
export interface ValueOpenAPI {
  /**
   * specification is the OpenAPI specification.
   */
  specification?: OpenAPIV3_1.Document;

  /**
   * routes are the HTTP routes of the OpenAPI specification.
   */
  routes?: Route[];

  /**
   * resources are the resources registered in the OpenAPI specification.
   */
  resources?: Class[];
}
