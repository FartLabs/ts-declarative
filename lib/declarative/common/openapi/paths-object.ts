import type { OpenAPIV3_1 } from "openapi-types";
import type { Class } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";

/**
 * pathsObjectOf returns the paths object of the OpenAPI class.
 */
export function pathsObjectOf<TClass extends Class>(
  target: TClass,
): OpenAPIV3_1.PathsObject | undefined {
  return getPrototypeValue<ValuePathsObject>(target)?.paths;
}

/**
 * reducePathsObject reduces the paths object of the OpenAPI
 * specification.
 */
export function reducePathsObject(
  paths: OpenAPIV3_1.PathsObject,
  pathsObject: OpenAPIV3_1.PathsObject,
): OpenAPIV3_1.PathsObject {
  for (const path in pathsObject) {
    paths[path] = { ...paths[path], ...pathsObject[path] };
  }

  return paths;
}

/**
 * ValuePathsObject is the value of the paths object of the OpenAPI
 * specification.
 */
export interface ValuePathsObject {
  /**
   * paths is the paths object of the OpenAPI specification.
   */
  paths?: OpenAPIV3_1.PathsObject;
}
