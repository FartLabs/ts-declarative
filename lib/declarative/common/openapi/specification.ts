import type { OpenAPIV3_1 } from "openapi-types";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { getPrototypeValue } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import { jsonSchemaOf } from "#/lib/declarative/common/json-schema/json-schema.ts";
import { pathsObjectOf, reducePathsObject } from "./paths-object.ts";

/**
 * specificationOf returns the OpenAPI specification of the class.
 */
export function specificationOf<TClass extends Class>(
  target: TClass,
): OpenAPIV3_1.Document | undefined {
  return getPrototypeValue<ValueOpenAPISpecification>(target)?.specification;
}

/**
 * openapiSpec is the decorator for OpenAPI specification.
 */
export const openapiSpec: (
  options?: OpenAPISpecificationDecoratorOptions,
) => (target: Class) => Class = createOpenAPISpecificationDecoratorFactory();

/**
 * createOpenAPISpecificationDecoratorFactory is the factory function for the
 * OpenAPI decorator.
 */
export function createOpenAPISpecificationDecoratorFactory(): (
  options?: OpenAPISpecificationDecoratorOptions | undefined,
) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (options?: OpenAPISpecificationDecoratorOptions) => {
      return [declarativeOpenAPISpecification(options)];
    },
  });
}

/**
 * OpenAPISpecificationDecoratorOptions is the options for the OpenAPI
 * specification decorator.
 */
export interface OpenAPISpecificationDecoratorOptions {
  /**
   * specification is the base OpenAPI specification.
   */
  specification?: Partial<OpenAPIV3_1.Document>;

  /**
   * resources are the resources of the OpenAPI specification.
   */
  resources?: Class[];
}

/**
 * defaultOpenAPI is the default OpenAPI version.
 */
export const defaultOpenAPI = "3.0.1";

/**
 * defaultOpenAPIInfoVersion is the default OpenAPI info version.
 */
export const defaultOpenAPIInfoVersion = "1.0.0";

/**
 * declarativeOpenAPISpecification is the declarative function for OpenAPI specification.
 */
export function declarativeOpenAPISpecification<
  TValue extends ValueOpenAPISpecification,
>(
  options?: OpenAPISpecificationDecoratorOptions,
): Declarative<TValue> {
  return (value, name) => {
    return {
      ...value,
      specification: {
        ...options?.specification,
        openapi: options?.specification?.openapi ?? defaultOpenAPI,
        info: {
          ...options?.specification?.info,
          title: options?.specification?.info?.title ?? name,
          version: options?.specification?.info?.version ??
            defaultOpenAPIInfoVersion,
        },
        paths: {
          ...options?.specification?.paths,
          ...(options?.resources
            ?.map((resource) => {
              const pathsObject = pathsObjectOf(resource);
              if (pathsObject === undefined) {
                throw new Error(`pathsObject is required for ${resource.name}`);
              }

              return pathsObject;
            })
            .reduce(reducePathsObject, {}) ?? {}),
        },
        components: {
          ...options?.specification?.components,
          schemas: {
            ...options?.specification?.components?.schemas,
            ...Object.fromEntries(
              options?.resources?.map((resource) => {
                const jsonSchema = jsonSchemaOf(resource);
                if (jsonSchema === undefined) {
                  throw new Error(
                    `jsonSchema is required for ${resource.name}`,
                  );
                }

                return [resource.name, jsonSchema];
              }) ?? [],
            ),
          },
        },
      },
    } as TValue;
  };
}

/**
 * ValueOpenAPISpecification is the value associated with an OpenAPI specification.
 */
export interface ValueOpenAPISpecification {
  /**
   * specification is the top-level object of the OpenAPI specification.
   */
  specification?: OpenAPIV3_1.Document;
}
