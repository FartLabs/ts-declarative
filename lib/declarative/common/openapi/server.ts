import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";
import type {
  RouterDecoratorOptions,
  ValueRouterRoutes,
} from "#/lib/declarative/common/router/router.ts";
import { declarativeRoutes } from "#/lib/declarative/common/router/router.ts";
import type {
  OpenAPISpecificationDecoratorOptions,
  ValueOpenAPISpecification,
} from "./specification.ts";
import { declarativeOpenAPISpecification } from "./specification.ts";

/**
 * openapi is the decorator for OpenAPI server.
 */
export const openapi: (
  options?: OpenAPIServerDecoratorOptions | undefined,
) => (target: Class) => Class = createOpenAPIServerDecoratorFactory();

/**
 * createOpenAPIServerDecoratorFactory is the factory function for the
 * OpenAPI server decorator.
 */
export function createOpenAPIServerDecoratorFactory(): (
  options?: OpenAPIServerDecoratorOptions | undefined,
) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (
      options?: OpenAPIServerDecoratorOptions,
    ): Array<Declarative<ValueOpenAPIServer>> => {
      return [
        declarativeOpenAPISpecification(options),
        declarativeRoutes(options),
      ];
    },
  });
}

/**
 * OpenAPIServerDecoratorOptions is the options for the OpenAPI server
 * decorator.
 */
export interface OpenAPIServerDecoratorOptions
  extends OpenAPISpecificationDecoratorOptions, RouterDecoratorOptions {}

/**
 * ValueOpenAPIServer is the value associated with an OpenAPI server.
 */
export interface ValueOpenAPIServer
  extends ValueOpenAPISpecification, ValueRouterRoutes {}
