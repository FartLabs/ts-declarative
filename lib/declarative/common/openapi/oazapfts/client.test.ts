import { assertSnapshot } from "@std/testing/snapshot";
import { Project } from "ts-morph";
import { createTypeInfoDecoratorFactory } from "#/lib/declarative/common/type-info/type-info.ts";
import { jsonSchema } from "#/lib/declarative/common/json-schema/json-schema.ts";
import { openapi } from "#/lib/declarative/common/openapi/openapi.ts";
import {
  standardCreate,
  standardGet,
} from "#/lib/declarative/common/google-aip/methods/mod.ts";
import { generateOazapftsClientOf } from "./client.ts";

const typeInfo = createTypeInfoDecoratorFactory(
  new Project(),
  "./lib/declarative/common/openapi/oazapfts/client.test.ts",
);

@standardCreate()
@standardGet()
@jsonSchema()
@typeInfo()
class Person {
  public constructor(public name: string) {}
}

@openapi({
  specification: {
    openapi: "3.0.1",
    info: { title: "App", version: "0.0.1" },
    components: {},
  },
  resources: [Person],
})
class App {}

Deno.test("generateOazapftsClientOf generate client from openapi specification", async (t) => {
  const sourceCode = await generateOazapftsClientOf(App);
  await assertSnapshot(t, sourceCode);
});
