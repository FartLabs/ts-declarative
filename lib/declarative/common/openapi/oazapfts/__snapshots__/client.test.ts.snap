export const snapshot = {};

snapshot[`generateOazapftsClientOf generates client from openapi specification 1`] = `
'/**
 * App
 * 0.0.1
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
export type Person = {
    name: string;
};
/**
 * Gets Person
 */
export function getPeopleByName(name: string, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: Person;
    }>(\`/people/\${encodeURIComponent(name)}\`, {
        ...opts
    }));
}
/**
 * Creates Person
 */
export function postPeople(person: Person, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: Person;
    }>("/people", oazapfts.json({
        ...opts,
        method: "POST",
        body: person
    })));
}
'
`;
